### PR TITLE
[inductor] share cse cache during vectorized indirect load

### PR DESCRIPTION
**Pull request by zhuhaozhe**
Fix https://github.com/pytorch/pytorch/issues/123502

`swap_buffer` in not needed in vectorized indirect load, remove it to share cse buffer.
```
auto tmp8 =
[&]
{
    __at_align__ std::array<int64_t, 16> tmpbuf;
    tmp7.store(tmpbuf.data());
    return tmpbuf;
}
()
;
//
// other codes
//
// also store tmp7 here (redundant tmp16)
auto tmp16 =
[&]
{
    __at_align__ std::array<int64_t, 16> tmpbuf;
    tmp7.store(tmpbuf.data());
    return tmpbuf;
}
()
;
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124921
* __->__ #124597

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -2318,7 +2318,7 @@ def vec_to_array(vec_var: CppCSEVariable) -> CppCSEVariable:
             assert vec_var.is_vec
             code = BracesBuffer()
             code.writeline("[&]")
-            with self.swap_buffers(code), code.indent():
+            with self.swap_buffers(code, cse_cache=self.cse.cache), code.indent():
```
[the remark this answers is not in the document]
Comment (zhuhaozhe) [reply torch/_inductor/codegen/cpp.py:2321]: > When new variables are defined in the scope, they cannot be used outside the scope afterwards. So, it is not safe to reuse the existing `self.cse.cache` here. I guess we want to create a wrapper around the cache abstraction that supports scoping, allowing variables and expressions defined within a scope to be valid only within that scope while still leveraging the existing cache from the outer scope before it. How about we create something like below?
> 
> ```python
> def scope_cse(cse):
>     new_cse = cse.clone()
>     new_cse.cache = ScopedDict(cse.cache)
>     new_cse.reduction_cache = ScopedDict(cse.reduction_cache)
>     new_cse.store_cache = ScopedDict(cse.store_cache)
>     return new_cse
> ```
> 
> Here the `ScopedDict` allows read-only access to the original dict while uses a separate dict to hold newly-added items:
> 
> ```python
> class ScopedDict:
>     def __init__(self, original_dict):
>         self.original_dict = original_dict
>         self.new_items = {}
>     
>     def __getitem__(self, key):
>         if key in self.new_items:
>             return self.new_items[key]
>         return self.original_dict[key]
>     
>     def __setitem__(self, key, value):
>         self.new_items[key] = value
>     
>     def __contains__(self, key):
>         return key in self.new_items or key in self.original_dict
>     
>     def get(self, key, default=None):
>         if key in self.new_items:
>             return self.new_items[key]
>         return self.original_dict.get(key, default)
> ```

Sounds better, I will try this way.

**Comment by zhuhaozhe** [comment]
Hi, @jgong5. For this issue https://github.com/pytorch/pytorch/issues/123502, your recommendation for `ScopedDict` should not work because
```
            auto tmp8 =
            [&]
            {
                __at_align__ std::array<int64_t, 16> tmpbuf;
                tmp7.store(tmpbuf.data());
                return tmpbuf;
            }
            ()
            ;
            auto tmp9 =
            [&]
            {
                __at_align__ std::array<float, 16> tmpbuf;
                #pragma GCC unroll 16
                for (long x0_inner = 0; x0_inner < 16; x0_inner++)
                {
                    tmpbuf[x0_inner] = out_ptr0[static_cast<long>(tmp8[x0_inner])];
                }
                return at::vec::Vectorized<float>::loadu(tmpbuf.data());
            }
            ()
            ;
```
the `tmp8` is created inside this [scope](https://github.com/pytorch/pytorch/blob/58806d65317c1a890d536c0c347598bfb1ee337f/torch/_inductor/codegen/cpp.py#L2360) [here](https://github.com/pytorch/pytorch/blob/58806d65317c1a890d536c0c347598bfb1ee337f/torch/_inductor/codegen/cpp.py#L2352). But actually I observed the `tmp8` should be able to be used outside because the `buffer` here in `csevar = self.cse.generate(buffer, code)` is not inside the lambda function (of `tmp9`) scope.
```
For my current understanding we may not need to always put a `swap_buffer` after a `[&]`, for `_load_or_store_non_contiguous` ,should it be safe to remove `swap buffer` here?

**Comment by zhuhaozhe** [comment]
As for the `ScopedDict`, it indeed help.
I found a case `test_AllenaiLongformerBase_repro_cpu` and make at as a UT in another PR https://github.com/pytorch/pytorch/pull/124921.
It will only generate `tmp_x=static_cast<int>(256)` once after `ScopedDict` impl, and it will be 3 times before it.

**Review by jgong5 — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**